oag/src/main/java/org/owasp/oag/controllers/JwksController.java

```diff
@@ -0,0 +1,95 @@
+    private static final String JWKS_BASE_URI = "/.well-known/jwks";
+
+    private static final Logger log = LoggerFactory.getLogger(JwksController.class);
+
+    private final JwkStore jwkStore;
+    private final MainConfig config;
+
+    @Autowired
+    JwksController(MainConfig config, KeyManagementFactory keyManagementFactory) {
+        this.config = config;
+        jwkStore = keyManagementFactory.getJWKStore();
+    }
+
+    /**
+     * Provides all the JWT signing keys used by OAG in the format of a JWK Set specified at https://tools.ietf.org/html/rfc7517.
+     * Namely this means:
+     * <code>
+     * {"keys":
+     * [
+     * {"kty":"RSA",
+     * "n": "0vx7agoebGcQSuuPiLJXZptN9nndrQmbXEps2aiAFbWhM78LhWx
+     * 4cbbfAAtVT86zwu1RK7aPFFxuhDR1L6tSoc_BJECPebWKRXjBZCiFV4n3oknjhMs
+     * tn64tZ_2W-5JsGY4Hc5n9yBXArwl93lqt7_RN5w6Cf0h4QyQ5v-65YGjQR0_FDW2
+     * QvzqY368QQMicAtaSqzs8KJZgnYb9c7d0zgdAZHzu6qMQvRL5hajrn1n91CbOpbI
+     * SD08qNLyrdkt-bFTWhAI4vMQFh6WeZu0fM4lFd2NcRwr3XPksINHaQ-G_xBniIqb
+     * w0Ls1jF44-csFCur-kEgU8awapJzKnqDKgw",
+     * "e":"AQAB",
+     * "alg":"RS256",
+     * "use": "sig",
+     * "kid":"gdfgrgergt-fzrfzgf-4343-utft"}
+     * },
+     * {...}
+     * ]
+     * }
+     * </code>
+     *
+     * @return all currently valid OAG used JWT signing keys as JWK Set. Returns an empty array when there are no such keys.
+     */
+    @GetMapping(JWKS_BASE_URI)
+    public Mono<ResponseEntity<Map<String, ?>>> getAllSigningKeys() {
+        return ReactiveUtils.runBlockingProcedure(() -> {
+            LoggingUtils.contextual(() -> log.debug("Retrieving all available JWK ..."));
+            JWKSet keys = jwkStore.getSigningPublicKeys();
+            LoggingUtils.contextual(() -> log.debug("signing Keys successfully retrieved..."));
+            // note: JWKSet keys is never null and always return a proper empty array when no keys are there.
+            return keys.toJSONObject();
+        })
+                .map(response -> {
+                    return ResponseEntity.ok(response);
+                });
+    }
+
+    /**
+     * @param kid the key id of the JWT signing key requested.
+     * @return the OAG JWT signing key with passed in kid (key id) (as single entry in the JWK Set array). Returns an empty array when not found.
+     */
+    @GetMapping(JWKS_BASE_URI + "/{kid}")
+    public Mono<ResponseEntity<Map<String, ?>>> getSigningKeys(@PathVariable("kid") String kid) {
+        return ReactiveUtils.runBlockingProcedure(() -> {
+            LoggingUtils.contextual(() -> log.debug("Retrieving singing JWK with kid {}.", kid));
+            JWKSet key = jwkStore.getSigningPublicKey(kid);
+            LoggingUtils.contextual(() -> log.debug("signing Key {} successfully retrieved: {}.", kid, key.getKeyByKeyId(kid) != null));
+            return key.toJSONObject();
+        })
+                .map(response -> {
+                    return ResponseEntity.ok(response);
+                });
```

Comment:
runBlockingProcedure() is quite expensive. It would be better if we cache the result of the getSigningPublicKeys()  and serve them instantly if they are loaded multiple times. Or maybe we should keep the public keys always in memory, then we would have no blocking calls at all.